# 整数溢出

在电脑领域里所发生的溢出条件是, 运行单项数值计算时, 当计算产生出来的结果是非常大的, 大于寄存器或存储器所能存储或表示的能力限制就会发生溢出.

在不同的编程语言中, 对待溢出通常有以下几种不同的做法:

- 崩溃: 当溢出被侦测到时, 程序立即退出运行
- 忽略: 这是最普遍的作法, 忽略任何算数溢出

对于溢出的处理方法, Rust 在 debug 与 release 模式下是不同的. 在 debug 模式下编译时, Rust 会检查整数溢出, 如果发生这种行为, 会导致程序在运行时终止并报出运行时错误. 而如果在 release 模式下编译时, Rust 不会对整数溢出进行检查.

要显式处理溢出, 可以使用标准库提供的一些 .overflowing_* 方法:

```rust
fn main() {
    let a: u32 = 4294967295;
    let b: u32 = 2;

    let (r, is_overflow) = a.overflowing_add(b);
    println!("r={} is_overflow={}", r, is_overflow);
}
```